### ISSUE TITLE
Static and non-static member function templates with the same parameter types are accepted

### ISSUE THREAD

**Issue by Fedr** [llvm/llvm-project#52951]
This code
```
struct A {
    static int f(auto) { return 1; }
    int f(auto) requires true { return 2; }
};
```
is invalid per https://timsong-cpp.github.io/cppwp/n4861/class.static.mfct#2:
> There shall not be a static and a non-static member function with the same name and the same parameter types ([over.load]).

But Clang accepts it without a warning (as some other compilers do), which later produces confusing errors. Related discussion: https://stackoverflow.com/q/70542265/7325599

**Comment by Endilll** [comment]
Referenced note was modified to exclude quoted sentence in [P1787R6 Declarations and where to find them](https://wg21.link/p1787r6). I'm not sure what is the normative wording for this.
We still don't issue any relevant diagnostics: https://godbolt.org/z/5abKvYG3n
But with `main` not commented out, we've been issuing an error about identical mangling those two functions have.
So this might be related to https://github.com/itanium-cxx-abi/cxx-abi/issues/24

**Comment by cor3ntin** [comment]
Not a bug
https://compiler-explorer.com/z/Tso6G5xMq 

> http://eel.is/c++draft/basic#scope.scope-4.3

Two function or function template declarations declare corresponding overloads if:
  - both declare functions with the same non-object-parameter-type-list,18 equivalent ([temp.over.link]) **trailing requires 
    clauses** (if any, except as specified in [temp.friend]), and ... 

@Endilll is correct this was changed by [P1787](https://wg21.link/p1787), and a whole bunch of subsequent papers pertaining to `deducing this`. but notably, P1787 was applied as a DR

Removing the require makes the definition correspond https://compiler-explorer.com/z/sY6hdTzoK

**Comment by shafik** [comment]
CC @zygoloid @hubert-reinterpretcast  should this be ambiguous? After reading [expr.unary p6](https://eel.is/c++draft/expr.unary#op-6) and [over.over p6](https://eel.is/c++draft/over.over#6) I think this should be valid but wanted confirmation.

**Comment by cor3ntin** [comment]
@shafik I believe that case is https://eel.is/c++draft/over.over#4 specifically

`y` selects the second overload because  the target type is a PMF, and `x` the first because the target type is a function pointer.
Not sure what is going on with mangling though

**Comment by zygoloid** [comment]
Fixed in trunk by 4b163e343cfa54c8d55c9da73c70d58f55ea9df2.
